From a37a7a8bee10ceb5e969a2423a353d534e21d272 Mon Sep 17 00:00:00 2001
From: Alain Volmat <alain.volmat@foss.st.com>
Date: Wed, 26 Mar 2025 09:45:27 +0100
Subject: [PATCH] media: stm32: dcmipp: add default histogram area settings in
 dcmipp-isp.c

Since histogram capture is enabled as soon as statistics are enabled, it is
necessary to setup a default histogram area otherwise this would lead to
a bad configuration error from the histogram block. Unfortunately it is
not possible to perform this at the time of start_streaming the statcap
since this would potentially overright a previously done configuration
by the isp-param output device. The same way it is not possible to do
that on the isp-param output device since it is not mandatory to use
in order to get statistics. For this reason perform that when configuring
the isp block and set a default configuration.

Change-Id: I553d77d66a4950ebf21d7a92eb285bcc7e5238fd
Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/446850
Tested-by: Alain VOLMAT <alain.volmat@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Philippe CORNU <philippe.cornu@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Alain VOLMAT <alain.volmat@st.com>
Domain-Review: Philippe CORNU <philippe.cornu@foss.st.com>
---
 .../st/stm32/stm32-dcmipp/dcmipp-isp.c        | 32 +++++++++++++++++++
 1 file changed, 32 insertions(+)

diff --git a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-isp.c b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-isp.c
index c9bc6dc25de1..cd5dae8add66 100644
--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-isp.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-isp.c
@@ -516,6 +516,35 @@ static void dcmipp_isp_config_decimation(struct dcmipp_isp_device *isp,
 	reg_write(isp, DCMIPP_P1DECR, decr);
 }
 
+/* Histogram block - only available starting from stm32mp21 */
+#define DCMIPP_P1HSCR			0x8b0
+/* 4 Comp / 64 bins per comp / 1 region / decimated by 2*/
+#define DCMIPP_P1HSCR_DEFAULT		0x08411000
+
+#define DCMIPP_P1HSSTR			0x8b4
+#define DCMIPP_P1HSSTR_START(x, y)	((x) | ((y) << 16))
+
+#define DCMIPP_P1HSSZR			0x8b8
+#define DCMIPP_P1HSSZR_SIZE(w, h)	((w) | ((h) << 16))
+static void dcmipp_isp_config_histo(struct dcmipp_isp_device *isp,
+				    struct v4l2_rect *compose)
+{
+	if (!isp->ved.dcmipp->pipe_cfg->has_histo)
+		return;
+
+	/*
+	 * Configure a default setting for histogram
+	 * In order to be able to capture statistics even if histogram settings
+	 * aren't yet done, perform initial settings of histogram with always
+	 * valid settings
+	 */
+	reg_write(isp, DCMIPP_P1HSCR, DCMIPP_P1HSCR_DEFAULT);
+	reg_write(isp, DCMIPP_P1HSSTR,
+		  DCMIPP_P1HSSTR_START(compose->width / 4, compose->height / 4));
+	reg_write(isp, DCMIPP_P1HSSZR,
+		  DCMIPP_P1HSSZR_SIZE(compose->width / 2, compose->height / 2));
+}
+
 static int dcmipp_isp_s_stream(struct v4l2_subdev *sd, int enable)
 {
 	struct dcmipp_isp_device *isp = v4l2_get_subdevdata(sd);
@@ -556,6 +585,9 @@ static int dcmipp_isp_s_stream(struct v4l2_subdev *sd, int enable)
 
 		/* Configure Demosaicing */
 		dcmipp_isp_config_demosaicing(isp, sink_fmt);
+
+		/* Configure default ISP Histo area */
+		dcmipp_isp_config_histo(isp, compose);
 	} else {
 		if (isp->usecnt > 1)
 			goto out;
-- 
2.50.1

